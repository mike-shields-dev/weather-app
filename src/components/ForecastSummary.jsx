import React from "react";
import PropTypes from "prop-types";
import styles from "../styles/ForecastSummary.module.css";
import WeatherIconComponent from "./WeatherIconComponent";
import weekdayFromUnixTime from "../helpers/weekdayFromUnixTime";

const ForecastSummary = ({
  date,
  description,
  handleForecastSelect,
  icon,
  temperature,
  selectedDate,
}) => {
  const isDateToday = new Date().getDay() === new Date(date).getDay();
  const isSelected = date === selectedDate;
  const modifierClass = styles["forecast-summary--selected"];
  return (
    <div
      className={`${styles["forecast-summary"]} ${
        isSelected ? modifierClass : null
      }`}
      data-testid="forecast-summary"
    >
      <h2
        className={styles["forecast-summary__date"]}
        data-testid="forecast-summary-date"
      >
        {isDateToday ? "Today" : weekdayFromUnixTime(date)}
      </h2>
      <WeatherIconComponent iconId={icon} />
      <p
        className="forecast-summary__description"
        data-testid="forecast-summary-description"
      >
        {description}
      </p>
      <p
        className="forecast-summary__temperature"
        data-testid="forecast-summary-temperature"
      >
        <span>{temperature.max}</span>
        Â°C
      </p>
      <button
        className="shadow"
        type="button"
        onClick={() => handleForecastSelect(date)}
      >
        More details...
      </button>
    </div>
  );
};

ForecastSummary.propTypes = {
  date: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  handleForecastSelect: PropTypes.func.isRequired,
  icon: PropTypes.string.isRequired,
  selectedDate: PropTypes.number.isRequired,
  temperature: PropTypes.shape({
    max: PropTypes.number,
    min: PropTypes.number,
  }).isRequired,
};

export default ForecastSummary;
