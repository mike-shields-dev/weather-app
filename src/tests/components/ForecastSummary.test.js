import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ForecastSummary from "../../components/ForecastSummary";

const validProps = {
  date: 1525046400000,
  temperature: {
    max: 11,
    min: 4,
  },
  wind: {
    speed: 13,
    direction: "s",
  },
  humidity: 30,
  description: "Clear",
  icon: "800",
  handleForecastSelect: jest.fn(),
};
const { description, temperature } = validProps;

describe("ForecastSummary", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<ForecastSummary {...validProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("displays correct day of week, or 'Today'", () => {
    const weekdayUnixTimeLookUp = {
      Sunday: 1662883154000,
      Monday: 1662969554000,
      Tuesday: 1663055954000,
      Wednesday: 1663142354000,
      Thursday: 1663228754000,
      Friday: 1663315154000,
      Saturday: 1663401554000,
    };

    Object.entries(weekdayUnixTimeLookUp).forEach(([weekday, unixTime]) => {
      const { getByText } = render(
        <ForecastSummary
          data-testid="forecast-summary"
          {...validProps}
          date={weekdayUnixTimeLookUp[weekday]}
        />
      );
      const isToday = new Date(unixTime).getDay() === new Date().getDay();

      if (isToday) {
        expect(getByText("Today")).toBeInTheDocument();
      } else {
        expect(getByText(weekday)).toBeInTheDocument();
      }
    });
  });

  it("displays description", () => {
    const { getByText } = render(
      <ForecastSummary data-testid="forecast-summary" {...validProps} />
    );

    expect(getByText(description)).toBeInTheDocument();
  });

  it("displays a weather icon", () => {
    const { getByTestId } = render(
      <ForecastSummary data-testid="forecast-summary" {...validProps} />
    );

    expect(getByTestId("weather-icon-component")).toBeInTheDocument();
  });

  it("displays temperature", () => {
    const { getByText } = render(
      <ForecastSummary data-testid="forecast-summary" {...validProps} />
    );

    expect(getByText(temperature.max)).toBeInTheDocument();
  });

  it("displays a 'More details...' button which calls handleForecastSelect callback when clicked", () => {
    const { getByText } = render(
      <ForecastSummary data-testid="forecast-summary" {...validProps} />
    );
    const detailsButton = getByText("More details...");

    fireEvent.click(detailsButton);

    expect(validProps.handleForecastSelect).toHaveBeenCalled();
  });
});
