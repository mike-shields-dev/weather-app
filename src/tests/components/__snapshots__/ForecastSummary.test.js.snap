// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastSummary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-summary null"
    data-testid="forecast-summary"
  >
    <h2
      class="forecast-summary__date"
      data-testid="forecast-summary-date"
    >
      Monday
    </h2>
    <div
      class="weather-icon-component"
      data-testid="weather-icon-component"
    >
      <i
        class="wi wi-day-sunny"
        name="owm"
        style="transform: scale(2.75);"
      />
    </div>
    <p
      class="forecast-summary__description"
      data-testid="forecast-summary-description"
    >
      Clear
    </p>
    <p
      class="forecast-summary__temperature"
      data-testid="forecast-summary-temperature"
    >
      <span>
        11
      </span>
      Â°C
    </p>
    <button
      class="shadow"
      type="button"
    >
      More details...
    </button>
  </div>
</DocumentFragment>
`;
